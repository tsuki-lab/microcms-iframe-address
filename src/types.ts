export type FormState = {
  /** 郵便番号 */
  postalCode: string
  /** 都道府県 */
  addressLevel1: string
  /** 市区町村 */
  addressLevel2: string
  /** 町名以下 */
  streetAddress: string
}
