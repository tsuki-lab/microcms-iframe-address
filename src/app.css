.example {
  display: inline-block;
  color: #68688a;
  font-size: 1.2rem;
  font-weight: normal;
  line-height: 1.4;
  padding-bottom: 8px;
  word-break: break-all;
}

.example::before {
  content: "ä¾‹: ";
}

.error {
  color: var(--color-error);
  font-size: 1.2rem;
}