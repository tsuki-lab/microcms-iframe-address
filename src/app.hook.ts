import { Reducer, useEffect, useReducer } from 'react'
import { useMicroCMSIframe } from './hooks/use-microcms-iframe'
import { useZipcloud } from './hooks/use-zipcloud'

const initFormState = {
  postalCode: '', // 郵便番号
  addressLevel1: '', // 都道府県
  addressLevel2: '', // 市区町村
  streetAddress: '', // 町名以下
}

type FormState = typeof initFormState

type FormReducer = Reducer<FormState, Partial<FormState>>

export const useApp = () => {
  const [formState, formDispatch] = useReducer<FormReducer>(
    (prev, action) => ({ ...prev, ...action }),
    initFormState
  )

  const [message, postHandler] = useMicroCMSIframe<FormState>({
    height: 510,
  })

  const { error: zipError, handler, loading } = useZipcloud()

  // microCMSに登録されている情報をStateに詰める
  useEffect(() => {
    if (!message?.data) return
    formDispatch(message.data)
  }, [message])

  /** 郵便番号から住所を検索する */
  const searchAddressByZip = async () => {
    const postalCode = formState.postalCode.replace(
      /^([0-9]{3})-?([0-9]{4})$/,
      '$1$2'
    )
    const { data: zips } = await handler(postalCode)
    if (!zips) return
    formDispatch({
      postalCode,
      addressLevel1: zips[0].address1,
      addressLevel2: zips[0].address2,
      streetAddress: zips[0].address3,
    })
  }

  // 郵便番号・住所・建物名のいずれかが更新されたらmicroCMSに送る
  useEffect(() => {
    const values = [
      formState.postalCode.replace(/^([0-9]{3})-?([0-9]{4})$/, '〒$1-$2'),
      formState.addressLevel1,
      formState.addressLevel2,
      formState.streetAddress,
    ]
    postHandler({
      description: values.filter(Boolean).join(' '),
      data: formState,
    })
  }, [formState, postHandler])

  return {
    formState,
    formDispatch,
    zipError,
    searchAddressByZip,
    loading,
  }
}
