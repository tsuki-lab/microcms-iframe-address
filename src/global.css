:root {
  --color: #563bff;
  --color-link: #563bff;
  --color-table: #563bff;
  --color-text: #21213b;
  --color-loader: #563bff;
  --color-error: #ff1f4d;
  --width-card-wide: 100%;
  font-size: 62.5%;

  /* Cancel MVP.css DarkMode */
  --color-accent: #118bee15;
  --color-bg: #fff;
  --color-bg-secondary: #e9e9e9;
  --color-secondary: #920de9;
  --color-secondary-accent: #920de90b;
  --color-shadow: #f4f4f4;
  --color-text-secondary: #999;
}

body {
  font-size: 1.6rem;
}

form {
  border: none;
  padding: 0 1px;
  box-shadow: none;
}

button[type='reset'] {
  color: var(--color);
  border: 1px solid var(--color);
  background-color: #fff;
}

/* microCMS input style */
input {
  padding: 16px;
  border: 1px solid #cdcde0;
  border-radius: 4px;
  box-sizing: border-box;
  background-color: #fff;
  min-width: 240px;
  height: 40px;
  font-size: 16px;
  box-shadow: 0 1px 4px rgb(0 0 0 / 8%);
}

select {
  padding: 8px 16px;
  border: 1px solid #ddd;
  border-radius: 4px;
  box-sizing: border-box;
  background-color: #fff;
  min-width: 240px;
  height: 40px;
  font-size: 16px;
  transition: box-shadow 0.2s ease;
  box-shadow: 0 1px 4px rgb(0 0 0 / 8%);
}

input:focus,
select:focus {
  outline: 1px solid #563bff;
}
