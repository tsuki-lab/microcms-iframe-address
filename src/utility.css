.w-full {
  width: 100%;
}

.red {
  color: var(--color-error);
}