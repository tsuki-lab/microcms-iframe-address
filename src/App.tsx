import { Reducer, useEffect, useMemo, useReducer } from 'react'
import { Loading } from './components/Loading'
import { useMicroCMSIframe } from './hooks/use-microcms-iframe'
import { useZipcloud } from './hooks/use-zipcloud/use-zipcloud'

import './app.css'

type FormState = {
  zip: string,
  address: string,
  building: string
}

type FormReducer = Reducer<FormState, Partial<FormState>>

function App() {
  const { state, postHandler } = useMicroCMSIframe<FormState>({
    height: 410
  })
  const { error: zipError, handler, loading } = useZipcloud()

  const isLoading = useMemo(() => loading, [loading])

  const [formState, formDispatch] = useReducer<FormReducer>((prev, action) => {
    return { ...prev, ...action }
  }, {
    zip: '',
    address: '',
    building: ''
  })

  const searchZip = async () => {
    const { data: zips } = await handler(formState.zip)
    if (zips) {
      formDispatch({address: `${zips[0].address1}${zips[0].address2}${zips[0].address3}`})
    }
  }

  useEffect(() => {
    if (state.defaultMessage?.data) {
      formDispatch(state.defaultMessage.data)
    }
  }, [state.defaultMessage])

  useEffect(() => {
    const values = [formState.zip.replace(/([0-9]{3})([0-9]{4})/, ''), formState.address, formState.building]
    postHandler({
      description: values.join(' '),
      data: formState
    })
  }, [formState])

  return (
    <>
      <form>
        <label htmlFor="zip">
          郵便番号<br />
          <span className='example'>1640001</span>
        </label>

        <input
          id="zip"
          type="text"
          required
          autoComplete='off'
          value={formState.zip}
          onChange={(e) => formDispatch({zip: e.target.value})}
        />
        <p className="error">{zipError}</p>
        <button type='button' onClick={searchZip}>郵便番号で住所を検索</button>

        <label htmlFor="address">
          住所<br />
          <span className='example'>東京都中野区中野5-65-5</span>
        </label>
        <input
          id="address"
          type="text"
          className="w-full"
          required
          autoComplete='off'
          value={formState.address}
          onChange={(e) => formDispatch({address: e.target.value})}
        />

        <label htmlFor="building">
          建物名<br />
          <span className='example'>豊島興業ビル7F</span>
        </label>
        <input
          id="building"
          type="text"
          className="w-full"
          autoComplete='off'
          value={formState.building}
          onChange={(e) => formDispatch({building: e.target.value})}
        />
      </form>
      {isLoading && <Loading />}
    </>
  )
}

export default App
