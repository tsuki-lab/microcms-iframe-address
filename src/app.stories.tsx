import { expect } from '@storybook/jest';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { userEvent, screen, waitFor } from '@storybook/testing-library';

import { App } from './App';

export default {
  title: 'App',
  component: App,
} as ComponentMeta<typeof App>;

const Template: ComponentStory<typeof App> = (args) => <App {...args} />;

export const Initial = Template.bind({});
Initial.storyName = '初回描画';
Initial.play = async () => {
  await waitFor(
    async () => {
      expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
    },
    { timeout: 5000 }
  );
};

export const InputPostalCode = Template.bind({});
InputPostalCode.storyName = '郵便番号の入力';
InputPostalCode.play = async (ctx) => {
  await Initial.play?.(ctx);
  await userEvent.type(screen.getByLabelText('郵便番号'), '1640001', {
    delay: 50,
  });
  expect(screen.getByLabelText('郵便番号')).toHaveValue('1640001');
};

export const ComplementAddress = Template.bind({});
ComplementAddress.storyName = '住所検索結果';
ComplementAddress.play = async (ctx) => {
  await InputPostalCode.play?.(ctx);
  userEvent.click(screen.getByRole('button', { name: '郵便番号で住所を検索' }));
  expect(await screen.findByRole('progressbar')).toBeInTheDocument();
  await waitFor(
    async () => {
      expect(screen.queryByRole('progressbar')).not.toBeInTheDocument();
    },
    { timeout: 5000 }
  );
  [
    { label: '郵便番号', value: '1640001' },
    { label: '都道府県', value: '東京都' },
    { label: '市区町村', value: '中野区' },
    { label: '町名以下', value: '中野' },
  ].forEach(({ label, value }) => {
    expect(screen.getByLabelText(label)).toHaveValue(value);
  });
};

export const ErrorNotPostalCodeValue = Template.bind({});
ErrorNotPostalCodeValue.storyName = 'Error:郵便番号未入力';
ErrorNotPostalCodeValue.play = async (ctx) => {
  await Initial.play?.(ctx);
  userEvent.click(screen.getByRole('button', { name: '郵便番号で住所を検索' }));
  expect(await screen.findByRole('alert')).toHaveTextContent(
    '郵便番号が入力されていません。'
  );
};
