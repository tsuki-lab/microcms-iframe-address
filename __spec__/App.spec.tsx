import { render } from '@testing-library/react'
import { describe, expect, test } from 'vitest'
import { App } from '../src/App'

describe('testing App.tsx', () => {
  test('snapshot testing', () => {
    const { asFragment } = render(<App />)
    expect(asFragment()).toMatchSnapshot()
  })
})
