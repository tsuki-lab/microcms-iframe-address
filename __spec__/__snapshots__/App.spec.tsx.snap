// Vitest Snapshot v1

exports[`testing App.tsx > snapshot testing 1`] = `
<DocumentFragment>
  <form
    autocomplete="off"
  >
    <label
      for="postal-code"
    >
      郵便番号
      <br />
      <span
        class="example"
      >
        1640001
      </span>
    </label>
    <input
      defaultvalue=""
      id="postal-code"
      name="postal-code"
      type="text"
    />
    <p
      class="error"
    />
    <button
      type="button"
    >
      郵便番号で住所を検索
    </button>
    <label
      for="address-level1"
    >
      都道府県
      <br />
      <span
        class="example"
      >
        東京都
      </span>
    </label>
    <input
      defaultvalue=""
      id="address-level1"
      type="text"
    />
    <label
      for="address-level2"
    >
      市区町村
      <br />
      <span
        class="example"
      >
        中野区
      </span>
    </label>
    <input
      defaultvalue=""
      id="address-level2"
      type="text"
    />
    <label
      for="street-address"
    >
      町名以下
      <br />
      <span
        class="example"
      >
        中野5-65-5 豊島興業ビル7F
      </span>
    </label>
    <input
      class="w-full"
      defaultvalue=""
      id="street-address"
      type="text"
    />
  </form>
</DocumentFragment>
`;
